internal/engine/upper_test.go

```diff
@@ -1661,6 +1675,95 @@ func TestPodContainerStatus(t *testing.T) {
 	f.assertAllBuildsConsumed()
 }
 
+func TestPodAddedToStateOrNotByTemplateHash(t *testing.T) {
+	deployedHash := k8s.PodTemplateSpecHash("some-hash-abc")
+	nonMatchingHash := k8s.PodTemplateSpecHash("danger-will-robinson")
+	ancestorUID := types.UID("some-ancestor")
+	podID := k8s.PodID("special-pod")
+	mName := model.ManifestName("foobar")
+
+	tests := []struct {
+		name                      string
+		ptshMatch                 bool // does the pod template spec hash of incoming pod event match the PTSH we most recently deployed?
+		ancestorSeen              bool // does the k8s runtime state already know about the ancestor UID of our pod?
+		withExistingPod           bool // instantiate runtime state with an old version of the pod we're getting events for
+		expectUpdatePodOnManifest bool // expect pod event to cause an update to manifestState.Pods
+	}{
+		{"new ancestor + ptsh OK", true, false, false, true},
+		{"new ancestor + ptsh not OK", false, false, false, false},
+		{"existing ancestor/new pod + ptsh OK", true, true, false, true},
+		{"existing ancestor/new pod + ptsh not OK", false, true, false, false},
+		{"existing pod + ptsh OK", true, true, true, true},
+		{"existing pod + ptsh not OK", false, true, true, true},
+	}
+
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			f := newTestFixture(t)
+			defer f.TearDown()
+			manifest := f.newManifest(mName.String())
+
+			f.b.nextDeployedUID = ancestorUID
+			f.b.nextPodTemplateSpecHashes = []k8s.PodTemplateSpecHash{deployedHash}
+			f.Start([]model.Manifest{manifest}, true)
+
+			_ = f.nextCallComplete()
+
+			// set up runtime state
+			st := f.store.LockMutableStateForTesting()
+			ms, ok := st.ManifestState(model.ManifestName(mName))
+			require.True(t, ok, "couldn't find manifest state for %s", mName)
+
+			runtime := ms.GetOrCreateK8sRuntimeState()
+			runtime.Pods = make(map[k8s.PodID]*store.Pod)
+			if test.ancestorSeen {
+				runtime.PodAncestorUID = ancestorUID
+			}
+
+			if test.withExistingPod {
+				runtime.Pods[podID] = &store.Pod{
+					PodID:  podID,
+					Status: "Running",
+				}
+			}
+			ms.RuntimeState = runtime
+			f.store.UnlockMutableState()
+
+			// Dispatch pod event
+			podHash := deployedHash
+			if !test.ptshMatch {
+				podHash = nonMatchingHash
+			}
+			pod := podbuilder.New(f.T(), manifest).
+				WithPodID(podID.String()).
```

Comment:
it seems weird to always use the same podID here as for the preexisting pod - do you think it's worth testing that dimension?

(I could imagine an implementation of the pod matching logic where we have a good old pod and then a bad new pod comes in and incorrectly wipes out the good old pod, when really the good old pod should be kept, but I think regardless of our behavior in that situation, the way the app works, that case will never arise?)

Reply:
I mean the case I'm testing is where we already have information about the pod, and we get in a new event about the same pod. good point about the additional dimension tho, let me see how much work it'd be to add that.